File: 692. Top K Frequent Words/Python/Solution.py
class Solution:
    def topKFrequent(self, words, k):
        """
        :type words: List[str]
        :type k: int
        :rtype: List[str]
        """
        counts = {}
        for word in words:
            if word not in counts:
                counts[word] = 0
            counts[word] += 1
        heap = []
        for word, count in counts.items():
            heapq.heappush(heap, (-count, word))
        result = []
        for i in range(k):
            if not heap:
                break
            result.append(heapq.heappop(heap)[1])
        return result

File: 692. Top K Frequent Words/Java/Solution.java
class Solution {
    public List<String> topKFrequent(String[] words, int k) {
        class Item implements Comparable<Item> {
            String word;
            int count;

            Item(String word, int count) {
                this.word = word;
                this.count = count;
            }

            @Override
            public int compareTo(Item o) {
                if (count == o.count) {
                    return o.word.compareTo(word);
                }
                return count - o.count;
            }
        }
        Map<String, Integer> counts = new HashMap<>();
        for (String word : words) {
            if (!counts.containsKey(word)) {
                counts.put(word, 0);
            }
            counts.put(word, counts.get(word) + 1);
        }
        PriorityQueue<Item> heap = new PriorityQueue<>();
        for (Map.Entry<String, Integer> entry : counts.entrySet()) {
            heap.add(new Item(entry.getKey(), entry.getValue()));
            if (heap.size() > k) {
                heap.poll();
            }
        }
        List<String> result = new ArrayList<>();
        while (!heap.isEmpty()) {
            result.add(heap.poll().word);
        }
        Collections.reverse(result);
        return result;
    }
}